internal/distributed/connection_manager.go

```diff
@@ -0,0 +1,309 @@
+package distributed
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	grpc_opentracing "github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing"
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/types"
+	"github.com/milvus-io/milvus/internal/util/retry"
+	"github.com/milvus-io/milvus/internal/util/sessionutil"
+	"github.com/milvus-io/milvus/internal/util/trace"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"go.uber.org/zap"
+)
+
+type ConnectionManager struct {
+	session *sessionutil.Session
+
+	dependencies map[string]struct{}
+
+	rootCoord  types.RootCoord
+	queryCoord types.QueryCoord
+	dataCoord  types.DataCoord
+	indexCoord types.IndexCoord
+	queryNodes map[string]types.QueryNode
+	dataNodes  map[string]types.DataNode
+	indexNodes map[string]types.IndexNode
+
+	taskMu     sync.RWMutex
+	buildTasks map[int64]*buildClientTask
+	notify     chan int64
+
+	closeCh chan struct{}
+}
+
+func NewConnectionManager(session *sessionutil.Session) *ConnectionManager {
+	return &ConnectionManager{
+		session: session,
+
+		dependencies: make(map[string]struct{}),
+
+		queryNodes: make(map[string]types.QueryNode),
+		dataNodes:  make(map[string]types.DataNode),
+		indexNodes: make(map[string]types.IndexNode),
+
+		buildTasks: make(map[int64]*buildClientTask),
+	}
+}
+
+func (cm *ConnectionManager) AddDependency(serverName string) error {
```

Comment:
serverName is inaccurate, serverName usually indicates one querynode/datanode name.  here role or rolename make more sense to me